arch/rv32i/src/support.rs

```diff
@@ -23,16 +23,20 @@ pub unsafe fn atomic<F, R>(f: F) -> R
 where
     F: FnOnce() -> R,
 {
-    let original_mstatus = CSR.mstatus.extract();
-    if original_mstatus.is_set(mstatus::mie) {
-        CSR.mstatus
-            .modify_no_read(original_mstatus, mstatus::mie::CLEAR);
-    }
+    // Read the mstatus MIE field and disable machine mode interrupts
+    // atomically
+    let original_mie = CSR.mstatus.read_and_clear_field(mstatus::mie);
+
+    // Machine mode interrupts are disabled, execute the atomic
+    // (uninterruptible) function
     let res = f();
-    if original_mstatus.is_set(mstatus::mie) {
-        CSR.mstatus
-            .modify_no_read(original_mstatus, mstatus::mie::SET);
+
+    // If the mstatus MIE field was set prior to clearing it, reenable
+    // machine mode interrupts
+    if original_mie != 0 {
+        CSR.mstatus.read_and_set_field(mstatus::mie);
```

Comment:
We could also potentially get rid of this branch, by just using the `original_mie` shifted by mstatus::mie.shift as the bitmask to `read_and_set_bits` directly. Is that worth it, or should we keep using the `read_and_set_field` here (which looks a bit nicer)?